File: csharp/2015/day01b/Program.cs
using System;
using System.IO;

class Program
{
    static void Usage()
    {
        Console.Error.WriteLine("usage: dotnet run <input file>");
        Environment.Exit(1);
    }

    static int Process(string filename)
    {
        int counter = 0;
        int pos = 0;

        try
        {
            using ( StreamReader reader = new StreamReader(filename) )
            {
                string? line;
                while ( (line = reader.ReadLine()) != null )
                {
                    foreach ( char ch in line )
                    {
                        pos++;
                        switch ( ch )
                        {
                            case '(':
                                counter++;
                                break;
                            case ')':
                                counter--;
                                break;
                        }
                        if ( counter < 0 ) {
                            return pos;
                        }
                    }
                }
            }
        }
        catch ( IOException e )
        {
            Console.Error.WriteLine("Error reading file: " + e.Message);
            Environment.Exit(1);
        }

        return 0;
    }

    static void Main(string[] args)
    {
        if ( args.Length < 1 )
        {
            Usage();
        }

        string filename = args[0];
        int result = Process(filename);
        Console.WriteLine("result = " + result);
    }
}

File: java/2015/day01b/Program.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Program {
    public static void usage() {
        System.err.println("usage: java -jar Program.jar <input file>");
        System.exit(1);
    }

    public static int process(String filename) {
        int counter = 0;
        int pos = 1;
        try ( BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ( (line = reader.readLine()) != null ) {
                for ( char ch : line.toCharArray() ) {
                    switch ( ch ) {
                        case '(':
                            counter++;
                            break;
                        case ')':
                            counter--;
                            break;
                    }
                    if ( counter < 0 ) {
                        return pos;
                    } else {
                        pos++;
                    }
                }
            }
        } catch ( IOException e ) {
            System.err.println("Error reading file: " + e.getMessage());
            System.exit(1);
        }
        return 0;
    }

    public static final void main(String[] args) {
        if ( args.length < 1 ) {
            usage();
        }
        String filename = args[0];
        int result = process(filename);
        System.out.println("result = " + result);
    }
}
